File: go/wafer-sdk-go/services/crypto.go
package services

import (
	"encoding/json"
	"fmt"
	"strconv"

	wafer "github.com/wafer-run/wafer-run/go/wafer-sdk-go"
	"github.com/wafer-run/wafer-run/go/wafer-sdk-go/gen/wafer/crypto"
)

// --- CallBlock-based implementations (context-aware) ---

// CryptoHashCtx computes a hash of the given password using CallBlock.
func CryptoHashCtx(ctx wafer.Context, password string) (string, error) {
	msg := wafer.NewMessage("crypto.hash", []byte(password))
	result := ctx.CallBlock("wafer-run/crypto", msg)
	if result.Error != nil {
		return "", fmt.Errorf("%s: %s", result.Error.Code, result.Error.Message)
	}
	return string(result.Response.Data), nil
}

// CryptoCompareHashCtx compares a plaintext password against a hash using
// CallBlock. Returns nil on match, error on mismatch.
func CryptoCompareHashCtx(ctx wafer.Context, password, hash string) error {
	msg := wafer.NewMessage("crypto.compare_hash", []byte(password))
	msg.SetMeta("hash", hash)
	result := ctx.CallBlock("wafer-run/crypto", msg)
	if result.Error != nil {
		return fmt.Errorf("%s: %s", result.Error.Code, result.Error.Message)
	}
	return nil
}

// CryptoSignCtx creates a signed JWT token from claims JSON with the given
// expiry in seconds using CallBlock.
func CryptoSignCtx(ctx wafer.Context, claims string, expirySecs uint64) (string, error) {
	msg := wafer.NewMessage("crypto.sign", []byte(claims))
	msg.SetMeta("expiry_secs", strconv.FormatUint(expirySecs, 10))
	result := ctx.CallBlock("wafer-run/crypto", msg)
	if result.Error != nil {
		return "", fmt.Errorf("%s: %s", result.Error.Code, result.Error.Message)
	}
	return string(result.Response.Data), nil
}

// CryptoVerifyCtx verifies a JWT token and returns the claims JSON string
// using CallBlock.
func CryptoVerifyCtx(ctx wafer.Context, token string) (string, error) {
	msg := wafer.NewMessage("crypto.verify", []byte(token))
	result := ctx.CallBlock("wafer-run/crypto", msg)
	if result.Error != nil {
		return "", fmt.Errorf("%s: %s", result.Error.Code, result.Error.Message)
	}
	return string(result.Response.Data), nil
}

// CryptoRandomBytesCtx generates n cryptographically secure random bytes
// using CallBlock.
func CryptoRandomBytesCtx(ctx wafer.Context, n uint32) ([]byte, error) {
	msg := wafer.NewMessage("crypto.random_bytes", nil)
	msg.SetMeta("n", strconv.FormatUint(uint64(n), 10))
	result := ctx.CallBlock("wafer-run/crypto", msg)
	if result.Error != nil {
		return nil, fmt.Errorf("%s: %s", result.Error.Code, result.Error.Message)
	}
	var bytes []byte
	if err := json.Unmarshal(result.Response.Data, &bytes); err != nil {
		// If JSON unmarshal fails, try using raw data (may be raw bytes)
		return result.Response.Data, nil
	}
	return bytes, nil
}

// --- Legacy direct-import implementations (backward compatible) ---

// CryptoHash computes a hash of the given password (e.g., bcrypt).
// When CallBlock is available, it routes through the "wafer-run/crypto" block.
func CryptoHash(password string) (string, error) {
	if wafer.HasCallBlock() {
		return CryptoHashCtx(wafer.NewContext(), password)
	}
	return crypto.Hash(password)
}

// CryptoCompareHash compares a plaintext password against a hash.
// Returns nil on match, error on mismatch.
func CryptoCompareHash(password, hash string) error {
	if wafer.HasCallBlock() {
		return CryptoCompareHashCtx(wafer.NewContext(), password, hash)
	}
	return crypto.CompareHash(password, hash)
}

// CryptoSign creates a signed JWT token from claims JSON with the given
// expiry in seconds.
func CryptoSign(claims string, expirySecs uint64) (string, error) {
	if wafer.HasCallBlock() {
		return CryptoSignCtx(wafer.NewContext(), claims, expirySecs)
	}
	return crypto.Sign(claims, expirySecs)
}

// CryptoVerify verifies a JWT token and returns the claims JSON string.
func CryptoVerify(token string) (string, error) {
	if wafer.HasCallBlock() {
		return CryptoVerifyCtx(wafer.NewContext(), token)
	}
	return crypto.Verify(token)
}

// CryptoRandomBytes generates n cryptographically secure random bytes.
func CryptoRandomBytes(n uint32) ([]byte, error) {
	if wafer.HasCallBlock() {
		return CryptoRandomBytesCtx(wafer.NewContext(), n)
	}
	return crypto.RandomBytes(n)
}

File: crates/wafer-sdk-rs/src/services/crypto.rs
//! Crypto service client — calls `wafer/crypto` block via `call-block`.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::wafer::block_world::runtime;
use crate::wafer::block_world::types::{Action, Message};

/// Crypto error type.
#[derive(Debug, Clone)]
pub struct CryptoError {
    pub kind: String,
    pub message: String,
}

impl std::fmt::Display for CryptoError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.kind, self.message)
    }
}

impl std::error::Error for CryptoError {}

// --- Internal request/response types ---

#[derive(Serialize)]
struct HashReq<'a> {
    password: &'a str,
}

#[derive(Serialize)]
struct CompareHashReq<'a> {
    password: &'a str,
    hash: &'a str,
}

#[derive(Serialize)]
struct SignReq<'a> {
    claims: &'a str,
    expiry_secs: u64,
}

#[derive(Serialize)]
struct VerifyReq<'a> {
    token: &'a str,
}

#[derive(Serialize)]
struct RandomBytesReq {
    n: u32,
}

#[derive(Deserialize)]
struct StringResp {
    value: String,
}

#[derive(Deserialize)]
struct BytesResp {
    data: Vec<u8>,
}

// --- Helpers ---

fn make_msg(kind: &str, data: &impl Serialize) -> Message {
    Message {
        kind: kind.to_string(),
        data: serde_json::to_vec(data).unwrap_or_default(),
        meta: Vec::new(),
    }
}

fn call_crypto(msg: &Message) -> Result<Vec<u8>, CryptoError> {
    let result = runtime::call_block("wafer/crypto", msg);
    match result.action {
        Action::Error => {
            let err_msg = result.error
                .map(|e| e.message)
                .unwrap_or_else(|| "unknown crypto error".to_string());
            let kind = if err_msg.contains("mismatch") {
                "password_mismatch"
            } else if err_msg.contains("hash") {
                "hash_error"
            } else if err_msg.contains("sign") {
                "sign_error"
            } else if err_msg.contains("verify") || err_msg.contains("token") {
                "verify_error"
            } else {
                "other"
            };
            Err(CryptoError { kind: kind.into(), message: err_msg })
        }
        _ => Ok(result.response.map(|r| r.data).unwrap_or_default()),
    }
}

fn call_crypto_parse<T: serde::de::DeserializeOwned>(msg: &Message) -> Result<T, CryptoError> {
    let data = call_crypto(msg)?;
    serde_json::from_slice(&data).map_err(|e| CryptoError {
        kind: "other".into(),
        message: format!("failed to parse response: {e}"),
    })
}

// --- Public API ---

/// Produce a one-way hash of a password.
pub fn hash(password: &str) -> Result<String, CryptoError> {
    let msg = make_msg("crypto.hash", &HashReq { password });
    let resp: StringResp = call_crypto_parse(&msg)?;
    Ok(resp.value)
}

/// Check a password against a hash. Returns Ok(()) if match.
pub fn compare_hash(password: &str, hash: &str) -> Result<(), CryptoError> {
    let msg = make_msg("crypto.compare_hash", &CompareHashReq { password, hash });
    call_crypto(&msg)?;
    Ok(())
}

/// Create a signed token from claims with the given expiry in seconds.
pub fn sign(claims: &HashMap<String, serde_json::Value>, expiry_secs: u64) -> Result<String, CryptoError> {
    let claims_json = serde_json::to_string(claims).unwrap_or_default();
    let msg = make_msg("crypto.sign", &SignReq { claims: &claims_json, expiry_secs });
    let resp: StringResp = call_crypto_parse(&msg)?;
    Ok(resp.value)
}

/// Verify a token and return its claims.
pub fn verify(token: &str) -> Result<HashMap<String, serde_json::Value>, CryptoError> {
    let msg = make_msg("crypto.verify", &VerifyReq { token });
    let data = call_crypto(&msg)?;
    // The response is the claims JSON directly
    serde_json::from_slice(&data).map_err(|e| CryptoError {
        kind: "other".into(),
        message: format!("failed to parse claims: {e}"),
    })
}

/// Generate n cryptographically-secure random bytes.
pub fn random_bytes(n: u32) -> Result<Vec<u8>, CryptoError> {
    let msg = make_msg("crypto.random_bytes", &RandomBytesReq { n });
    let resp: BytesResp = call_crypto_parse(&msg)?;
    Ok(resp.data)
}
